File: src/algorithms/hash-maps/sliding-window/longest-substring-without-repeating/LongestSubstringWithoutRepeating_test.cpp
#include "sources/LongestSubstringWithoutRepeating.cpp"
#include <cassert>
#include <iostream>

int main() {
    assert(longestSubstringWithoutRepeating("abcabcbb") == 3);
    assert(longestSubstringWithoutRepeating("bbbbb") == 1);
    assert(longestSubstringWithoutRepeating("pwwkew") == 3);
    assert(longestSubstringWithoutRepeating("") == 0);
    assert(longestSubstringWithoutRepeating("a") == 1);
    assert(longestSubstringWithoutRepeating("abcde") == 5);
    assert(longestSubstringWithoutRepeating("abba") == 2);
    assert(longestSubstringWithoutRepeating("dvdf") == 3);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: src/algorithms/hash-maps/sliding-window/longest-substring-without-repeating/__tests__/longest-substring-without-repeating_test.rs
include!("../sources/longest-substring-without-repeating.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_returns_3_for_abcabcbb() {
        assert_eq!(longest_substring_without_repeating("abcabcbb"), 3);
    }

    #[test]
    fn test_returns_1_for_bbbbb() {
        assert_eq!(longest_substring_without_repeating("bbbbb"), 1);
    }

    #[test]
    fn test_returns_3_for_pwwkew() {
        assert_eq!(longest_substring_without_repeating("pwwkew"), 3);
    }

    #[test]
    fn test_returns_0_for_empty_string() {
        assert_eq!(longest_substring_without_repeating(""), 0);
    }

    #[test]
    fn test_returns_1_for_single_char() {
        assert_eq!(longest_substring_without_repeating("a"), 1);
    }

    #[test]
    fn test_returns_5_for_abcde() {
        assert_eq!(longest_substring_without_repeating("abcde"), 5);
    }

    #[test]
    fn test_returns_2_for_abba() {
        assert_eq!(longest_substring_without_repeating("abba"), 2);
    }

    #[test]
    fn test_returns_3_for_dvdf() {
        assert_eq!(longest_substring_without_repeating("dvdf"), 3);
    }
}
